### DISCUSSION TITLE
MIP solver choice 2023

### DISCUSSION THREAD

**Discussion by CADBIMDeveloper** [google/or-tools#3969] [General] [answered]
Hi experts!

I'm ready to start a new project with several optimization capabilities, I've prepared all the math back-end for it. 
It's a completely MIP problem and it doesn't require any non-linear support. 
I can use CP-SAT (if I scale the problem by 10 I can consider everything as integers), math_opt (by the way, what is the status of it in 2023? 2 years ago it was "experimental" according to https://github.com/google/or-tools/discussions/2935#discussioncomment-1711669), SCIP, CBC, GLPK, whatever else... 

What do you think is the best choice?

**Comment by lesshaste** [comment]
Don't forget https://highs.dev/ .

**Reply by lperron** [reply]
not yet supported.

**Comment by lperron** [comment, marked as answer]
Commercial: gurobi, xpress, copt (not yet supported by OR-Tools), no particular order
Non commercial: CP-SAT preferred if you can run more than 8 cores and if the problem does not need continuous variables, SCIP otherwise. Highs when supported.

**Reply by CADBIMDeveloper** [reply]
Thank you!

**Comment by gregy4** [comment]
I recommend to model it in or-tools as mip problem. You can directly see results in solvers supported by or-tools including cp-sat when you choose SAT solver. Export to lp format allows you to check results for other solvers for example on neos-server.org.
For my latest mip problem cp-sat solver was much faster compared to all available commercial solvers so I finally rewrite my problem to cp-sat model due to access to callbacks. Performance of pure cp-sat solver and mip mapping of cp-sat solver were similar.

**Reply by CADBIMDeveloper** [reply]
Hi, thank you! Yes, I'm familiar with it and pretty happy about cp-sat performance. My concerns were about the support of different non-linear capabilities of the CP-SAT which could have (or don't have) a price on the performance. My current task is a pure MIP, so I was interested in taking into account another options.

**Reply by gregy4** [reply]
I see, I think cp-sat solver checks characteristics of a problem during presolve phase (for example when only boolean variables are used) so you probably don't need to worry about it. But it works differently compared to mip solvers and tries various approaches in parallel via search workers.

With mip problem modeled in or-tools you can probably very shortly see whether some mip solver or cp-sat solver is noticeably better for your problem with one code (no need to model it twice).
